File: src/main/java/at/scheinecker/intellij/coco/CocoColorSettingsPage.java
package at.scheinecker.intellij.coco;

import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighter;
import com.intellij.openapi.options.colors.AttributesDescriptor;
import com.intellij.openapi.options.colors.ColorDescriptor;
import com.intellij.openapi.options.colors.ColorSettingsPage;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.Map;

/**
 * @author Thomas Scheinecker <a href="mailto:tscheinecker@gmail.com">tscheinecker@gmail.com</a>
 */
public class CocoColorSettingsPage implements ColorSettingsPage {
    private static final AttributesDescriptor[] DESCRIPTORS = new AttributesDescriptor[]{
            new AttributesDescriptor("Block Comment", CocoSyntaxHighlighter.BLOCK_COMMENT),
            new AttributesDescriptor("Line Comment", CocoSyntaxHighlighter.LINE_COMMENT),
            new AttributesDescriptor("Keyword", CocoSyntaxHighlighter.KEYWORD),
            new AttributesDescriptor("Identifier", CocoSyntaxHighlighter.IDENTIFIER),
            new AttributesDescriptor("Markup Tags", CocoSyntaxHighlighter.MARKUP_TAG),
            new AttributesDescriptor("Char", CocoSyntaxHighlighter.CHAR),
            new AttributesDescriptor("String", CocoSyntaxHighlighter.STRING),
            new AttributesDescriptor("Braces", CocoSyntaxHighlighter.BRACES),
            new AttributesDescriptor("Brackets", CocoSyntaxHighlighter.BRACKETS),
            new AttributesDescriptor("Parentheses", CocoSyntaxHighlighter.PARENTHESES),
            new AttributesDescriptor("Operations ( + | - | .. | = )", CocoSyntaxHighlighter.OPERATION_SIGN),
            new AttributesDescriptor("Terminal Symbol", CocoSyntaxHighlighter.TERMINATOR)
    };
    public static final String COLOR_SETTINGS_DEMO_ATG_FILE = "/at/scheinecker/intellij/coco/colorSettingsDemo.ATG";
    private static final String FALLBACK_SETTINGS_CONTENT = "// test\n" +
            "\n" +
            "/* ... */\n" +
            "\n" +
            "import\n" +
            "    dddda.asdfd.asdf;\n" +
            "import dddae.ass.*;\n" +
            "import test.Easdl;\n" +
            "import test.ad;\n" +
            "\n" +
            "COMPILER dfd2ks32l\n" +
            "\n" +
            "asfd\n" +
            "\n" +
            "TOKEN_IGNORECASE\n" +
            "// SCANNER SPECIFICATION\n" +
            "/*ddd*/\n" +
            "CHARACTERS\n" +
            "test = a + '\\u0AF0' .\n" +
            "somethingelse = a - 'c'.\n" +
            "another = 'a'..'z'.\n" +
            "\n" +
            "TOKENS\n" +
            "  number = daasdf.\n" +
            "  \"\\0\" =  'a' | de TOKEN_CONTEXT (\"b\\XXb\\\\b\\X\\vca\").\n" +
            "\n" +
            "PRAGMAS\n" +
            "ident = b . (. a a some st\n" +
            "uff .)\n" +
            "jident2 = a .\n" +
            "jetonemore = a . (. ad .)\n" +
            "\n" +
            "COMMENTS FROM \"a\" TO \"b\" NESTED\n" +
            "COMMENTS FROM \"/**\" TO eol\n" +
            "\n" +
            "IGNORE 'a'..'z'\n" +
            "\n" +
            "// PARSER SPECIFIATION\n" +
            "\n" +
            "TOKEN_PRODUCTIONS\n" +
            "ident < arbitrary text > (. arbitrary statements .) =\n" +
            "    WEAK eof < act attributes > | IF (ANY) { no } |\n" +
            "    IF (ANY) [\n" +
            "        IF (ANY) {\n" +
            "            IF (ANY) WEAK t\n" +
            "        }\n" +
            "    ] .\n" +
            "END ident .";

    @Nullable
    @Override
    public Icon getIcon() {
        return CocoIcons.FILE;
    }

    @NotNull
    @Override
    public SyntaxHighlighter getHighlighter() {
        return new CocoSyntaxHighlighter();
    }

    @NotNull
    @Override
    public String getDemoText() {
        final InputStream COLOR_SETTINGS_DEMO_ATG_FILE_INPUT_STREAM = CocoColorSettingsPage.class.getResourceAsStream(COLOR_SETTINGS_DEMO_ATG_FILE);
        if (COLOR_SETTINGS_DEMO_ATG_FILE_INPUT_STREAM == null) {
            return FALLBACK_SETTINGS_CONTENT;
        }
        URL myURL = this.getClass().getResource(COLOR_SETTINGS_DEMO_ATG_FILE);
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(COLOR_SETTINGS_DEMO_ATG_FILE_INPUT_STREAM))) {
            return reader.lines().sequential().reduce("", (a, b) -> a + "\n"+ b).substring(1);
        } catch (java.io.IOException e) {
            return FALLBACK_SETTINGS_CONTENT;
        }
    }

    @Nullable
    @Override
    public Map<String, TextAttributesKey> getAdditionalHighlightingTagToDescriptorMap() {
        return null;
    }

    @NotNull
    @Override
    public AttributesDescriptor[] getAttributeDescriptors() {
        return DESCRIPTORS;
    }

    @NotNull
    @Override
    public ColorDescriptor[] getColorDescriptors() {
        return new ColorDescriptor[0];
    }

    @NotNull
    @Override
    public String getDisplayName() {
        return "Cocol/R";
    }
}

File: src/main/java/at/scheinecker/intellij/coco/CocoColorSettingsPage.kt
package at.scheinecker.intellij.coco

import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighter
import com.intellij.openapi.options.colors.AttributesDescriptor
import com.intellij.openapi.options.colors.ColorDescriptor
import com.intellij.openapi.options.colors.ColorSettingsPage
import javax.swing.Icon

/**
 * @author Thomas Scheinecker [tscheinecker@gmail.com](mailto:tscheinecker@gmail.com)
 */
class CocoColorSettingsPage : ColorSettingsPage {

    override fun getIcon(): Icon? {
        return CocoIcons.FILE
    }

    override fun getHighlighter(): SyntaxHighlighter {
        return CocoSyntaxHighlighter()
    }

    override fun getDemoText(): String {
        return COLOR_SETTINGS_ATG_CONTENT
    }

    override fun getAdditionalHighlightingTagToDescriptorMap(): Map<String, TextAttributesKey>? {
        return null
    }

    override fun getAttributeDescriptors(): Array<AttributesDescriptor> {
        return DESCRIPTORS
    }

    override fun getColorDescriptors(): Array<ColorDescriptor> {
        return emptyArray()
    }

    override fun getDisplayName(): String {
        return "Cocol/R"
    }

    companion object {
        private val DESCRIPTORS = arrayOf(AttributesDescriptor("Block Comment", CocoSyntaxHighlighter.BLOCK_COMMENT), AttributesDescriptor("Line Comment", CocoSyntaxHighlighter.LINE_COMMENT), AttributesDescriptor("Keyword", CocoSyntaxHighlighter.KEYWORD), AttributesDescriptor("Identifier", CocoSyntaxHighlighter.IDENTIFIER), AttributesDescriptor("Markup Tags", CocoSyntaxHighlighter.MARKUP_TAG), AttributesDescriptor("Char", CocoSyntaxHighlighter.CHAR), AttributesDescriptor("String", CocoSyntaxHighlighter.STRING), AttributesDescriptor("Braces", CocoSyntaxHighlighter.BRACES), AttributesDescriptor("Brackets", CocoSyntaxHighlighter.BRACKETS), AttributesDescriptor("Parentheses", CocoSyntaxHighlighter.PARENTHESES), AttributesDescriptor("Operations ( + | - | .. | = )", CocoSyntaxHighlighter.OPERATION_SIGN), AttributesDescriptor("Terminal Symbol", CocoSyntaxHighlighter.TERMINATOR))
        private val COLOR_SETTINGS_ATG_CONTENT = """COMPILER Taste

Proc curProc;  // current program unit (procedure or main program)

CHARACTERS
  letter = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz".
  digit = "0123456789".
  cr  = '\r'.
  lf  = '\n'.
  tab = '\t'.

TOKENS
  ident  = letter {letter | digit}.
  number = digit {digit}.

COMMENTS FROM "/*" TO "*/" NESTED
COMMENTS FROM "//" TO lf

IGNORE cr + lf + tab

PRODUCTIONS

/*=================== Program and declarations ============================*/

Taste                    (. String name; .)
= "program"
  Ident<out name>        (. curProc = new Proc(name, null, this); .)
  "{"
  { VarDecl | ProcDecl }
  "}"                    (. for (Obj obj: curProc.locals) {
                            	if (obj instanceof Proc) ((Proc)obj).dump();
                            } .).
/*------------------------------------------------------------------------*/
VarDecl                  (. String name; Type type; .)
= Typ<out type>
  Ident<out name>        (. curProc.add(new Var(name, type)); .)
  { ',' Ident<out name>  (. curProc.add(new Var(name, type)); .)
  } ';'.
/*------------------------------------------------------------------------*/
Typ<out Type type>
=                        (. type = Type.INT; .)
 ( "int"
 | "bool"                (. type = Type.BOOL; .)
 ).
/*------------------------------------------------------------------------*/
ProcDecl                 (. String name; .)
= "void"
  Ident<out name>        (. Proc oldProc = curProc;
                            curProc = new Proc(name, oldProc, this);
                            oldProc.add(curProc); .)
  '(' ')'
  Block<out curProc.block> (. curProc = oldProc; .).


/*============================= Statements ===============================*/

Block<out Block b>       (. Stat s; .)
= '{'                    (. b = new Block(); .)
  { Stat<out s>          (. b.add(s); .)
  | VarDecl
  }
  '}'
  .
/*------------------------------------------------------------------------*/
Stat<out Stat s>         (. String name; Expr e; Stat s2; Block b; .)
=                        (. s = null; .)
( Ident<out name>        (. Obj obj = curProc.find(name); .)
  ( '='
		Expr<out e> ';'      (. s = new Assignment(obj, e); .)
	| '(' ')' ';'          (. s = new Call(obj); .)
	)

| "if"
	'(' Expr<out e> ')'
	Stat<out s>            (. s = new If(e, s); .)
	[ "else" Stat<out s2>  (. s = new IfElse(s, s2); .)
	]

| "while"
	'(' Expr<out e> ')'
	Stat<out s>            (. s = new While(e, s); .)

| "read"
	Ident<out name> ';'    (. s = new Read(curProc.find(name)); .)

| "write"
	Expr<out e> ';'        (. s = new Write(e); .)

| Block<out b>           (. s = b; .)
).



/*============================ Expressions ===============================*/

Expr<out Expr e>         (. Operator op; Expr e2; .)
= SimExpr<out e>
  [ RelOp<out op>
    SimExpr<out e2>      (. e = new BinExpr(e, op, e2); .)
  ].
/*------------------------------------------------------------------------*/
SimExpr<out Expr e>      (. Operator op; Expr e2; .)
= Term<out e>
  { AddOp<out op>
    Term<out e2>         (. e = new BinExpr(e, op, e2); .)
	}.
/*------------------------------------------------------------------------*/
Term<out Expr e>         (. Operator op; Expr e2; .)
= Factor<out e>
  { MulOp<out op>
    Factor<out e2>       (. e = new BinExpr(e, op, e2); .)
	}.
/*------------------------------------------------------------------------*/
Factor<out Expr e>       (. String name; .)
=                        (. e = null; .)
  ( Ident<out name>      (. e = new Ident(curProc.find(name)); .)
  | number               (. e = new IntCon(Integer.parseInt(t.val)); .)
  | '-' Factor<out e>    (. e = new UnaryExpr(Operator.SUB, e); .)
  | "true"               (. e = new BoolCon(true); .)
  | "false"              (. e = new BoolCon(false); .)
  ).
/*------------------------------------------------------------------------*/
Ident<out String name>
= ident                  (. name = t.val; .).
/*------------------------------------------------------------------------*/
AddOp<out Operator op>
=                        (. op = Operator.ADD; .)
  ( '+'
  | '-'                  (. op = Operator.SUB; .)
  ).
/*------------------------------------------------------------------------*/
MulOp<out Operator op>
=                        (. op = Operator.MUL; .)
  ( '*'
  | '/'                  (. op = Operator.DIV; .)
  ).
/*------------------------------------------------------------------------*/
RelOp<out Operator op>
=                        (. op = Operator.EQU; .)
  ( "=="
  | '<'                  (. op = Operator.LSS; .)
  | '>'                  (. op = Operator.GTR; .)
  ).

END Taste.
"""
    }
}
